File: src/job_status.py
"""In-memory status registry for report requests.

The registry intentionally does not survive an application restart. It is a
small, UI-facing step before report execution moves to a durable queue.
"""

from datetime import datetime
from threading import Lock
from typing import Literal, TypedDict

Status = Literal["queued", "running", "failed"]


class Job(TypedDict):
    criado_em: str
    data_fim: str
    data_inicio: str
    estado: str
    id_requisicao: str
    mensagem: str | None
    status: Status
    tipo: str


_jobs: dict[str, Job] = {}
_lock = Lock()


def register(
    request_id: str,
    report_type: str,
    states: list[str],
    start_date: str,
    end_date: str,
) -> None:
    with _lock:
        _jobs[request_id] = Job(
            criado_em=datetime.now().isoformat(timespec="seconds"),
            data_fim=end_date,
            data_inicio=start_date,
            estado=" · ".join(states),
            id_requisicao=request_id,
            mensagem=None,
            status="queued",
            tipo=report_type,
        )


def mark_running(request_id: str) -> None:
    with _lock:
        if request_id in _jobs:
            _jobs[request_id]["status"] = "running"


def mark_failed(request_id: str) -> None:
    with _lock:
        if request_id in _jobs:
            _jobs[request_id]["status"] = "failed"
            _jobs[request_id]["mensagem"] = "Não foi possível gerar este relatório."


def mark_succeeded(request_id: str) -> None:
    with _lock:
        _jobs.pop(request_id, None)


def list_jobs() -> list[Job]:
    with _lock:
        return list(_jobs.values())

File: go/internal/jobstatus/jobstatus.go
// Package jobstatus mirrors src/job_status.py: an in-memory registry that
// intentionally does not survive a restart.
package jobstatus

import (
	"strings"
	"sync"
	"time"
)

type Status string

const (
	StatusQueued  Status = "queued"
	StatusRunning Status = "running"
	StatusFailed  Status = "failed"
)

type Job struct {
	CriadoEm     string  `json:"criado_em"`
	DataFim      string  `json:"data_fim"`
	DataInicio   string  `json:"data_inicio"`
	Estado       string  `json:"estado"`
	IDRequisicao string  `json:"id_requisicao"`
	Mensagem     *string `json:"mensagem"`
	Status       Status  `json:"status"`
	Tipo         string  `json:"tipo"`
}

type Registry struct {
	mu   sync.Mutex
	jobs map[string]Job
}

func NewRegistry() *Registry {
	return &Registry{jobs: make(map[string]Job)}
}

func nowIsoSeconds() string {
	return time.Now().UTC().Format("2006-01-02T15:04:05-07:00")
}

func (r *Registry) Register(requestID, reportType string, states []string, startDate, endDate string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.jobs[requestID] = Job{
		CriadoEm:     nowIsoSeconds(),
		DataFim:      endDate,
		DataInicio:   startDate,
		Estado:       strings.Join(states, " · "),
		IDRequisicao: requestID,
		Mensagem:     nil,
		Status:       StatusQueued,
		Tipo:         reportType,
	}
}

func (r *Registry) MarkRunning(requestID string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	if job, ok := r.jobs[requestID]; ok {
		job.Status = StatusRunning
		r.jobs[requestID] = job
	}
}

func (r *Registry) MarkFailed(requestID string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	if job, ok := r.jobs[requestID]; ok {
		message := "Não foi possível gerar este relatório."
		job.Status = StatusFailed
		job.Mensagem = &message
		r.jobs[requestID] = job
	}
}

func (r *Registry) MarkSucceeded(requestID string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	delete(r.jobs, requestID)
}

func (r *Registry) List() []Job {
	r.mu.Lock()
	defer r.mu.Unlock()
	out := make([]Job, 0, len(r.jobs))
	for _, job := range r.jobs {
		out = append(out, job)
	}
	return out
}
